### ISSUE TITLE
Step matching for strings appears broken in 2.1.0

### ISSUE THREAD

**Issue by lukaso** [jnicklas/turnip#179]
The problem lies somewhere in the placeholder code here: https://github.com/jnicklas/turnip/compare/v2.0.2...a7efda5ca050a96c876462ac5a4c64a13da5c23c

I noticed that the `@default` value is memoized in two different places in the same `placeholder.rb` file. I don't know if that's the problem, but it might be.

The behaviour I'm seeing is that:

```
Then I see the 'quote expired' page
```

Is matching step definition:

```
step "I see the ':page_name' page" do |page_name|

end
```

But `page_name` is 'quote'.

Additionally I have these two steps defined in this order:

```
step "I see the 'quote expired' page do

end

step "I see the ':page_name' page" do |page_name|

end
```

And get an ambiguous step error in 2.1.0 which I didn't get in 2.0.2:
```
     Failure/Error: Then I see the 'quote expired' page

     Turnip::Ambiguous:
       Ambiguous step definitions
         - "I see the 'quote expired' page" (...spec/journey/renewal_buy_online/renewal_buy_online_spec.rb:174:in `block (2 levels) in <top (required)>')
         - "I see the ':page_name' page" (...spec/journey/renewal_buy_online/renewal_buy_online_spec.rb:178:in `block (2 levels) in <top (required)>')
```

**Comment by gongo** [comment]
Hi @lukaso .

Thanks for issue and I could confirm this issue. Sorry for the trouble. 🙇 
But, I think of this issue **as specification** .

## The reason

Step like the following has been resolved as ambiguous from before.

```ruby
step "this is ambiguous" do
end

step "this is :ambiguous" do
end
```

https://github.com/jnicklas/turnip/blob/v2.1.0/examples/steps/steps.rb#L17-L21

So, I think a natural thing that issue steps are resolved as ambiguous.
In other words, I think the behavior of the past was unnatural.

## As a specification change 🙏 

Again, sorry for the trouble. But I think of this issue **as specification** .

**Comment by lukaso** [comment]
@gongo I understand about the ambiguous steps issue, and that that is as it should be. However, there is still a severe bug:

> The behaviour I'm seeing is that:

```
Then I see the 'quote expired' page
```

> Is matching step definition:

```
step "I see the ':page_name' page" do |page_name|

end
```

> In the step `page_name` equals `'quote'` but should equal `'quote expired'`.

Something about the way it is capturing the placeholder has changed and is broken.

**Comment by lukaso** [comment]
@gongo And thank you for looking at this and responding :)

**Comment by gongo** [comment]
> In the step page_name equals 'quote' but should equal 'quote expired'.
> Something about the way it is capturing the placeholder has changed and is broken.

Oh, you're right. this is bug :scream:
I will fix it soon 🙇

**Comment by gongo** [comment]
Hi @lukaso 

Sorry for my late reply.

Please try to test using [capturing_default-placeholder_asitis](https://github.com/jnicklas/turnip/tree/capturing_default-placeholder_asitis) branch. (see: #182)
If there is no problem, I will release it as v2.1.1

**Comment by lukaso** [comment]
Hello @gongo, this doesn't appear to be working exactly right. It looks like with strings with new lines in them it is adding extra quotes. But most tests are working.

Here's my test:
```feature
  Scenario Outline: Business address should be present and in a clean format
    Given a Business with business name "<business_name>", address line one "<address_line_1>", address line two "<address_line_2>", town "<town>" and postcode "<postcode>"
    Then the business address should display as "<business_address>"

  Examples:
    | business_name | address_line_1 | address_line_2 | town | postcode | business_address                                       |
    | Acme Widget   | 1 The Street   | The District   | Town | QR1 2ST  | Acme Widget\n1 The Street\nThe District\nTown\nQR1 2ST |
    |               | 1 The Street   | The District   | Town | QR1 2ST  | 1 The Street\nThe District\nTown\nQR1 2ST              |
    | Acme Widget   | 1 The Street   |                | Town | QR1 2ST  | Acme Widget\n1 The Street\nTown\nQR1 2ST               |
```

These steps are failing with:

```diff
     Failure/Error: expect(@document_data[:business_address]).to eq(business_address)

       expected: "\"Acme Widget\n1 The Street\nTown\nQR1 2ST\""
            got: "Acme Widget\n1 The Street\nTown\nQR1 2ST"

       (compared using ==)

       Diff:

       @@ -1,5 +1,5 @@
       -"Acme Widget
       +Acme Widget
        1 The Street
        Town
       -QR1 2ST"
       +QR1 2ST
```

**Comment by lukaso** [comment]
Actually, the original problem is still happening:

```
     Turnip::Ambiguous:
       Ambiguous step definitions
         - "I see the 'quote expired' page" (spec/journey/renewal_buy_online/renewal_buy_online_spec.rb:174:in `block (2 levels) in <top (required)>')
         - "I see the ':page_name' page" (spec/journey/renewal_buy_online/renewal_buy_online_spec.rb:178:in `block (2 levels) in <top (required)>')
```

**Comment by gongo** [comment]
I see.

>       expected: "\"Acme Widget\n1 The Street\nTown\nQR1 2ST\""
>            got: "Acme Widget\n1 The Street\nTown\nQR1 2ST"

This result is due to the effects of https://github.com/jnicklas/turnip/pull/182/commits/b96b8435b24a219c7d7a6c560389ea62a90bec66 .
This commit is necessary in order not to divide a string that have not been surrounded by the quote and contains a space.
So, I think as specification change.

```
    "Jhon Doe"    \ 
  "\"Jhon Doe\""  =>  "Jhon Doe"
   "'Jhon Doe'"   /
```

**Comment by gongo** [comment]
> So, I think as specification change.

Sorry, this behaviour is not specification change (The original specification in v2.0.x).
Does you add quote dare for v2.1.0's bug ?

**Comment by lukaso** [comment]
@gongo I haven't made any changes to deal with v2.1.0's bug. I have not been using 2.1.0 except to test your fix. There were not quote changes.

**Comment by gongo** [comment]
@lukaso OK. I will try to fix 🙇

**Comment by gongo** [comment]
@lukaso I have a question.
How did you define `Then the business ...` step?  https://github.com/jnicklas/turnip/issues/179#issuecomment-223922336

```ruby
step 'the business address should display as ":business_address"' do |address|
  # ...
end

# or

step 'the business address should display as :business_address' do |address|
  # ...
end

# or other?
```

**Comment by lukaso** [comment]
Here are the two relevant step definitions:

```ruby
step "a Policy with business name :business_name, address line one :address_line_1, address line two :address_line_2, town :town and postcode :postcode" do |business_name, address_line_1, address_line_2, town, postcode|

step "the business address should display as :business_address" do |business_address|
```

**Comment by gongo** [comment]
@lukaso Thanks for reply.
Please retry to test using latest capturing_default-placeholder_asitis branch 🙏

**Comment by lukaso** [comment]
@gongo This is now working correctly! Thank you for fixing it. Let me know when you release the new version.

**Comment by gongo** [comment]
@lukaso I've released v2.1.1 🎉
